File: 113/path-sum-ii.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} sum
 * @return {number[][]}
 */
var pathSum = function (root, sum) {
    let ans = [];
    let tmp = [];

    function dfs(root, sum) {
        if (!root)
            return;

        tmp.push(root.val);
        if (!root.left && !root.right) {
            if (root.val === sum) {
                let res = tmp.map(item => item);
                ans.push(res);
            }
            return;
        }
        if (root.left) {
            dfs(root.left, sum - root.val);
            tmp.pop();
        }
        if (root.right) {
            dfs(root.right, sum - root.val);
            tmp.pop();
        }
    }
    dfs(root, sum);
    return ans;
};

File: 113/path-sum-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def __init__(self):
        self.tmp = []
        self.ans = []

    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        self.dfs(root, sum)
        return self.ans

    def dfs(self, root, sum):
        if root is None:
            return

        self.tmp.append(root.val)
        if root.left is None and root.right is None:
            if root.val == sum:
                res = list(map(lambda x: x, self.tmp))
                self.ans.append(res)
            return

        if root.left is not None:
            self.dfs(root.left, sum - root.val)
            self.tmp = self.tmp[:len(self.tmp)-1]

        if root.right is not None:
            self.dfs(root.right, sum - root.val)
            self.tmp = self.tmp[:len(self.tmp)-1]
